tools/cephfs-mirror: eliminate redundant ceph_close() call

ifed01 opened the pull request:
Fixes: https://tracker.ceph.com/issues/68853

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

joscollin commented on src/tools/cephfs_mirror/PeerReplayer.cc:
```diff
@@ -120,7 +120,6 @@ int opendirat(MountRef mnt, int dirfd, const std::string &relpath, int flags,
 
   int fd = r;
   r = ceph_fdopendir(mnt, fd, dirp);
-  ceph_close(mnt, fd);
```
I think ceph_close is necessary here. We are getting the dirp using `ceph_fdopendir`. Any thoughts @vshankar ?
https://github.com/ceph/ceph/blob/main/src/client/Client.cc#L9175

ifed01 replied:
No, it's not. Please see https://github.com/ceph/ceph/pull/55619 which states:
"Based on posix specification, the fd passed to fdopendir() will be closed by closedir(). However CephFS client wasn't doing that. If the user opened a directory using ceph_openat(), for example, and then passed the returned fd to ceph_fdopendir(), the created Fh associated with the new open was never destroyed.
"

vshankar commented:
jenkins test api

vshankar commented:
@joscollin PTAL for review and approve if the change looks good.

joscollin commented:
@vshankar Will check soon.

vshankar commented:
> @vshankar Will check soon.

Anything on this @joscollin ?

joscollin commented:
> > @vshankar Will check soon.
> 
> Anything on this @joscollin ?

Currently reviewing it. The new changes looks good at the first look. Checking in detail.

joscollin commented on src/tools/cephfs_mirror/PeerReplayer.cc:
```diff
@@ -1388,6 +1373,21 @@ int PeerReplayer::do_synchronize(const std::string &dir_root, const Snapshot &cu
+  // if c_fd has been used in ceph_fdopendir call then
+  // there is no need to clost this fd manually.
+  if (close_current) {
+    ceph_close(m_local_mount, fh.c_fd);
```
I think we need to use `ceph_closedir` here instead, which uses `dir_result_t`.

sajibreadd replied:
Shouldn't this ceph_closedir already closing fh.c_fd if it reaches there while popping the stack? In that case shouldn't be the ceph_close call redundent?

sajibreadd replied:
Okay, but yeah, my thought was regarding if ceph_closedir fails.

ifed01 replied:
ceph_closedir() has the only way to return an error - when cmount isn't mounted, nad Client::closedir always returns 0:

extern "C" int ceph_closedir(struct ceph_mount_info *cmount, struct ceph_dir_result *dirp)
{
  if (!cmount->is_mounted())
    return -CEPHFS_ENOTCONN;
  return cmount->get_client()->closedir(reinterpret_cast<dir_result_t*>(dirp));
}

So we might want to just put ceph_assert() here to make sure we never getting the error..

joscollin commented on src/tools/cephfs_mirror/PeerReplayer.cc:
```diff
@@ -1388,6 +1373,21 @@ int PeerReplayer::do_synchronize(const std::string &dir_root, const Snapshot &cu
+  // its used to acquire an exclusive lock on remote dir_root.
+
+  // if c_fd has been used in ceph_fdopendir call then
+  // there is no need to clost this fd manually.
```
nit: s/clost/close

joscollin commented on src/tools/cephfs_mirror/PeerReplayer.cc:
```diff
@@ -1560,6 +1552,17 @@ int PeerReplayer::do_synchronize(const std::string &dir_root, const Snapshot &cu
+  // its used to acquire an exclusive lock on remote dir_root.
+
+  ceph_close(m_local_mount, fh.c_fd);
+  ceph_close(fh.p_mnt, fh.p_fd);
```
Aren't we already closing `fh.p_mnt` via `ceph_close_snapdiff`?

ifed01 commented on src/tools/cephfs_mirror/PeerReplayer.cc:
```diff
@@ -1408,9 +1408,16 @@ int PeerReplayer::do_synchronize(const std::string &dir_root, const Snapshot &cu
+  std::queue<SyncEntry> sync_queue;
+
+  //start with initial/default entry
+  std::string epath = ".", npath = "", nabs_path = "", nname = "";
```
see above, goto can't cross var declarations.

ifed01 replied:
Done

sajibreadd commented:
`ceph_closedir` the fd internally then why are we closing the fd in opendirat?

ifed01 commented on src/tools/cephfs_mirror/PeerReplayer.cc:
```diff
@@ -1221,28 +1221,19 @@ int PeerReplayer::sync_perms(const std::string& path) {
-
+  ceph_dir_result *tdirp;
+  struct ceph_statx tstx;
+  std::stack<SyncEntry> sync_stack;
```
Updated as per @joscollin proposal.

joscollin commented on src/tools/cephfs_mirror/PeerReplayer.cc:
```diff
@@ -120,6 +120,9 @@ int opendirat(MountRef mnt, int dirfd, const std::string &relpath, int flags,
   r = ceph_fdopendir(mnt, fd, dirp);
+  if (r < 0) {
+    ceph_close(mnt, fd);
+  }
```
That's good.

joscollin reviewed (approved):
Now, `dout` is only for success case. But not a big deal. LGTM.

joscollin reviewed (changes requested):
@ifed01 Oops, now there's a redundant commit https://github.com/ceph/ceph/pull/60667/commits/a0df949879b7c19e38b00d3c119b95c57ee3a75e. Could you please fix this?

ifed01 commented:
> @ifed01 Oops, now there's a redundant commit [a0df949](https://github.com/ceph/ceph/commit/a0df949879b7c19e38b00d3c119b95c57ee3a75e). Could you please fix this?

@joscollin - yeah, squashed commits around ceph_close() into a single one.
Added another trivial one to get rid off build warning in Client.cc. It's irrelevant to cephfs-mirror just want to piggy-back this PR to cut the path if you don't mind...

joscollin commented:
> > @ifed01 Oops, now there's a redundant commit [a0df949](https://github.com/ceph/ceph/commit/a0df949879b7c19e38b00d3c119b95c57ee3a75e). Could you please fix this?
> 
> @joscollin - yeah, squashed commits around ceph_close() into a single one. 

Thanks.

> Added another trivial one to get rid off build warning in Client.cc. It's irrelevant to cephfs-mirror just want to piggy-back this PR to cut the path if you don't mind...

We won't catch that in the fs:mirror tests, but it certainly needs QA. So could you please create a new PR for that fix?

ifed01 commented:
> 
> > Added another trivial one to get rid off build warning in Client.cc. It's irrelevant to cephfs-mirror just want to piggy-back this PR to cut the path if you don't mind...
> 
> We won't catch that in the fs:mirror tests, but it certainly needs QA. So could you please create a new PR for that fix?

Done

ifed01 commented:
jenkins test api

joscollin commented:
@rishabh-d-dave 
The QA has a failure:
https://pulpito.ceph.com/rishabh-2024-12-09_14:46:35-fs-wip-rishabh-testing-20241204.130505-testing-default-smithi/8026609/

This is a known issue, which is tracked here: https://tracker.ceph.com/issues/68567.  Could you please drop that failed test and QA again, so that we could get this PR in?

rishabh-d-dave commented:
> @rishabh-d-dave The QA has a failure: https://pulpito.ceph.com/rishabh-2024-12-09_14:46:35-fs-wip-rishabh-testing-20241204.130505-testing-default-smithi/8026609/
> 
> This is a known issue, which is tracked here: https://tracker.ceph.com/issues/68567. Could you please drop that failed test

Drop that test how?

> and QA again, so that we could get this PR in?

QA again why? We already did a full QA run, there are no other mirror related failures there.

joscollin commented:
> > @rishabh-d-dave The QA has a failure: https://pulpito.ceph.com/rishabh-2024-12-09_14:46:35-fs-wip-rishabh-testing-20241204.130505-testing-default-smithi/8026609/
> > This is a known issue, which is tracked here: https://tracker.ceph.com/issues/68567. Could you please drop that failed test
> 
> Drop that test how?
> 
> > and QA again, so that we could get this PR in?
> 
> QA again why? We already did a full QA run, there are no other mirror related failures there.

https://github.com/ceph/ceph/pull/61068
That's the fix for the failure. I'll take up the QA for this PR, as I have more fixes coming in.

joscollin commented:
@vshankar 
QA passed: https://pulpito.ceph.com/jcollin-2024-12-13_12:51:29-fs:mirror-wip-jcollin-testing-20241213.065332-main-distro-default-smithi/.

Please merge this PR.
